### PR TITLE
chore(dev): add AI-assisted component maturity evaluation skill

### PR DESCRIPTION
**Pull request by pront**
## Summary

Tracking Vector component maturity (stable/beta/deprecated) is currently manual and ad-hoc. This adds a Claude Code skill that automates a monthly evaluation — collecting signals from git, GitHub issues, CUE metadata, and test infrastructure — and writes a structured markdown report with promotion candidates, watch list, and full inventory.

The skill lives at `.claude/skills/vector-components-maturity-eval/SKILL.md` and is invoked via `/vector-components-maturity-eval`.

**Signals collected per component:**
- Open bugs via the `type:Bug` GitHub issue Type field (not the legacy label)
- Test quality: E2E against a live external service for sources/sinks; unit test coverage for transforms (transforms have no external dependencies by design)
- Component age from git history and config churn over 6 months (hand-written + generated CUE)
- Docs quality (AI judgment: complete / partial / minimal)

Bug severity is assessed qualitatively — minor bugs don't block promotion; data-loss or crash bugs do.

## Vector configuration

N/A — developer tooling only.

## How did you test this PR?

Ran the skill and generated the June 2026 report. Compared against the May 2026 report from the previous skill version to validate the signal improvements.

## Change Type

- [ ] Bug fix
- [ ] New feature
- [ ] Dependencies
- [x] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [x] No. A maintainer will apply the `no-changelog` label to this PR.

### CODE REVIEW

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  for f in website/cue/reference/components/${kind}/*.cue; do
+    name=$(basename "$f" .cue)
+    paths=()
+
+    add_path() {
+      [ -e "$1" ] && paths+=("$1")
+    }
+
+    add_path "src/${kind}/${name}.rs"
+    add_path "src/${kind}/${name}"
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+     -maxdepth 1 -name "*.cue" | sort
+
+# Integration test directories
+ls tests/integration/
+```
+
+---
+
+## Phase 2: Bulk Signal Collection
+
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md:56]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+      sinks/sematext_logs) add_path "src/sinks/sematext/logs.rs" ;;
+      sinks/sematext_metrics) add_path "src/sinks/sematext/metrics.rs" ;;
+      sinks/splunk_hec_logs) add_path "src/sinks/splunk_hec/logs" ;;
+      sinks/splunk_hec_metrics) add_path "src/sinks/splunk_hec/metrics" ;;
+    esac
+
+    if [ ${#paths[@]} -eq 0 ]; then
+      count=0
+    elif command -v rg >/dev/null 2>&1; then
+      count=$(rg -n "$test_attr_regex" "${paths[@]}" 2>/dev/null | wc -l | tr -d ' ')
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+For every component, assign one recommendation:
+
+| Rec | Meaning |
+| --- | --- |
+| **promote** | Beta → stable candidate |
+| **keep** | No change warranted |
+| **watch** | Stable with concerning signals |
+| **deprecate-candidate** | Little activity, superseded, or already deprecated in CUE |
+
+**Promote** (beta only): 0–1 open bugs AND (integration test OR coverage ≥ 50%) AND age > 4 months AND churn ≤ 5 commits AND docs at least `partial`.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Makes sense but we will stop using this tag from now on.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+done
+```
+
+### 2c. Config churn — commits to CUE file in last 6 months
+
+```bash
+for kind in sources transforms sinks; do
+  for f in website/cue/reference/components/${kind}/*.cue; do
+    name=$(basename "$f" .cue)
+    count=$(git log --since="6 months ago" --oneline -- "$f" 2>/dev/null | wc -l | tr -d ' ')
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in 991710739b.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+run_id=$(gh run list --workflow=coverage.yml --branch=master --status=success \
+  --limit=1 --json databaseId --jq '.[0].databaseId')
+
+[ -z "$run_id" ] && { echo "ERROR: no successful coverage run found on master" >&2; exit 1; }
+echo "Coverage run: $run_id"
+
+# Download the archive-merged-lcov-report artifact (produced by coverage-upload job)
+rm -rf /tmp/vector-coverage
+gh run download "$run_id" -n archive-merged-lcov-report --dir /tmp/vector-coverage \
+  || { echo "ERROR: archive-merged-lcov-report artifact missing — run may predate the artifact upload step" >&2; exit 1; }
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Should probably inspect .github/workflows/coverage.yml for an up-to-date artifact name

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+  /^SF:/          { file=$0; sub(/^SF:/, "", file); lf=0; lh=0 }
+  /^LF:/          { lf=$0; sub(/^LF:/, "", lf) }
+  /^LH:/          { lh=$0; sub(/^LH:/, "", lh) }
+  /^end_of_record/ { if (lf+0 > 0) printf "%s|%.1f\n", file, (lh/lf)*100 }
+' /tmp/vector-coverage/merged-lcov.info
+```
+
+**Mapping files to components:** For each component `<name>` of type `<kind>`, collect all LCOV lines whose path starts with `src/<kind>/<name>` (catches both `src/sources/kafka.rs` and `src/sources/kafka/`). Average the coverage percentages across those lines. Components with no matching path are marked `n/a`.
+
+This resolves shared-directory sources (e.g. `src/sources/prometheus/scrape.rs` → `prometheus_scrape`) and shared utilities (e.g. `src/sources/prometheus/parser.rs` → counted for all three prometheus sources) without any explicit path mapping.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+# Emit file|coverage_pct for every source file in the report
+awk '
+  /^SF:/          { file=$0; sub(/^SF:/, "", file); lf=0; lh=0 }
+  /^LF:/          { lf=$0; sub(/^LF:/, "", lf) }
+  /^LH:/          { lh=$0; sub(/^LH:/, "", lh) }
+  /^end_of_record/ { if (lf+0 > 0) printf "%s|%.1f\n", file, (lh/lf)*100 }
+' /tmp/vector-coverage/merged-lcov.info
+```
+
+**Mapping files to components:** For each component `<name>` of type `<kind>`, collect all LCOV lines whose path starts with `src/<kind>/<name>` (catches both `src/sources/kafka.rs` and `src/sources/kafka/`). Average the coverage percentages across those lines. Components with no matching path are marked `n/a`.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+## Phase 2: Bulk Signal Collection
+
+Use single shell loops to collect all signals at once — do not make one Bash call per component.
+
+### 2a. Open GitHub bugs
+
+Issues use the GitHub issue **Type** field. The type name is `Bug`.
+
+```bash
+gh issue list --state open --search "type:Bug" --json number,title,url,labels --limit 500 2>/dev/null
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+# Emit file|coverage_pct for every source file in the report
+awk '
+  /^SF:/          { file=$0; sub(/^SF:/, "", file); lf=0; lh=0 }
+  /^LF:/          { lf=$0; sub(/^LF:/, "", lf) }
+  /^LH:/          { lh=$0; sub(/^LH:/, "", lh) }
+  /^end_of_record/ { if (lf+0 > 0) printf "%s|%.1f\n", file, (lh/lf)*100 }
+' /tmp/vector-coverage/merged-lcov.info
+```
+
+**Mapping files to components:** For each component `<name>` of type `<kind>`, collect all LCOV lines whose path starts with `src/<kind>/<name>` (catches both `src/sources/kafka.rs` and `src/sources/kafka/`). Average the coverage percentages across those lines. Components with no matching path are marked `n/a`.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Already addressed, stale.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+- Bugs are identified by the GitHub issue **Type** field (`type:Bug` in search) — issues use the Type field, not labels
+- Working directory is the Vector repo root
+- **Coverage artifact**: the `archive-merged-lcov-report` artifact is produced by the `coverage-upload` job in `.github/workflows/coverage.yml` and retained for 90 days. Download with `gh run download <run_id> -n archive-merged-lcov-report --dir /tmp/vector-coverage`. Runs older than 90 days will not have the artifact.
+- **Coverage thresholds**: ≥50% = well covered, 20–50% = partial, <20% = low, n/a = no data
+
+**Parent/shared CUE files**: Some CUE files define shared configuration for families of components and have no `development` field of their own (children inherit it). Known true parent files (exclude from per-component inventory): `sinks/aws_cloudwatch.cue`, `sinks/datadog.cue`, `sinks/gcp.cue`, `sinks/humio.cue`, `sinks/influxdb.cue`, `sinks/sematext.cue`, `sinks/splunk_hec.cue`. `sinks/statsd.cue` and `sources/syslog.cue` are real components whose `development` value is inherited — read the child CUE files to resolve it and include them in the inventory.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ef72b1e4a2.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+---
+
+## Phase 3: Read CUE Files
+
+Read each component's CUE file in batches of 10–15 (parallel Read calls in a single response). Extract:
+
+- `development` value — `"stable"`, `"beta"`, or `"deprecated"`
+- Whether `how_it_works` has substantive prose. If it references a shared CUE object, read that referenced object and judge the resolved prose; shared populated docs count as substantive.
+- Whether `description` (top-level) is meaningful: at least two sentences explaining what the component does and when to use it
+- Whether there are non-trivial `examples` in the configuration section
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ef72b1e4a2.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+## Phase 2: Bulk Signal Collection
+
+Use single shell loops to collect all signals at once — do not make one Bash call per component.
+
+### 2a. Open GitHub bugs
+
+Issues use the GitHub issue **Type** field. The type name is `Bug`.
+
+```bash
+gh issue list --state open --search "type:Bug" --json number,title,url,labels --limit 500 2>/dev/null
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ef72b1e4a2.

[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ef72b1e4a2.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+For every component, assign one recommendation:
+
+| Rec | Meaning |
+| --- | --- |
+| **promote** | Beta → stable candidate |
+| **keep** | No change warranted |
+| **watch** | Stable with concerning signals |
+| **deprecate-candidate** | Little activity, superseded, or already deprecated in CUE |
+
+**Promote** (beta only): 0–1 open bugs AND (integration test OR coverage ≥ 50%) AND age > 4 months AND churn ≤ 5 commits AND docs at least `partial`.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in f1bbf26679. Replaced hard count with severity-based judgment — no severity labels exist in the repo so the evaluator uses issue title/description to distinguish minor from critical bugs.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+## Phase 2: Bulk Signal Collection
+
+Use single shell loops to collect all signals at once — do not make one Bash call per component.
+
+### 2a. Open GitHub bugs
+
+Issues use the GitHub issue **Type** field. The type name is `Bug`.
+
+```bash
+gh issue list -R vectordotdev/vector --state open --search "type:Bug" --json number,title,url,labels --limit 1000 2>/dev/null
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ef72b1e4a2. Added `body` and `createdAt` to the `gh issue list` fetch.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+| Rec | Meaning |
+| --- | --- |
+| **promote** | Beta → stable candidate |
+| **keep** | No change warranted |
+| **watch** | Stable with concerning signals |
+| **deprecate-candidate** | Little activity, superseded, or already deprecated in CUE |
+
+**Promote** (beta only): No critical/major open bugs AND E2E test tier ✓ or ~ AND age > 4 months AND churn ≤ 5 commits AND docs at least `partial`. Minor or cosmetic bugs do not block promotion — use judgment based on issue title and description. A single data-loss or crash bug is blocking; a docs typo or edge-case UX issue is not.
+
+**Watch** (stable only): ≥ 3 open bugs, OR churn > 10 commits (API instability), OR E2E tier ✗ (no tests at all).
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Addressed in ec71cd375b. Watch criterion now explicitly exempts transforms from the E2E tier ✗ trigger — only sources and sinks are flagged for missing integration tests.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+## Phase 2: Bulk Signal Collection
+
+Use single shell loops to collect all signals at once — do not make one Bash call per component.
+
+### 2a. Open GitHub bugs
+
+Issues use the GitHub issue **Type** field (not a label). The type name is `Bug`.
+
+```bash
+gh issue list --state open --search "type:Bug" --json number,title,url --limit 500 2>/dev/null
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — Phase 4 already uses label-only matching.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+## Reference
+
+- CUE files at `website/cue/reference/components/{sources,transforms,sinks}/` are authoritative (ignore `generated/` subdirs)
+- Source implementations: `src/sources/<name>.rs` or `src/sources/<name>/`, same pattern for sinks and transforms
+- `gh` is pre-authenticated for `vectordotdev/vector`
+- Bugs are identified by the GitHub issue **Type** field (`type:Bug` in search), not by label — the old `type: bug` label is deprecated
+- Working directory is the Vector repo root
+
+**Parent/shared CUE files**: Some CUE files define shared configuration for families of components and have no `development` field of their own (children inherit it). These will appear as "unknown" status when grepped. Known parent files: `sinks/aws_cloudwatch.cue`, `sinks/datadog.cue`, `sinks/gcp.cue`, `sinks/humio.cue`, `sinks/influxdb.cue`, `sinks/sematext.cue`, `sinks/splunk_hec.cue`, and possibly `sinks/statsd.cue`, `sources/syslog.cue`. Exclude these from per-component counts; note them separately.
```
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — statsd and syslog inheritance resolution is covered in the Reference section.

File: .claude/skills/vector-components-maturity-eval/SKILL.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — coverage/lcov phase was removed from the skill in an earlier revision.

File: .claude/skills/vector-components-maturity-eval/SKILL.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — coverage/lcov phase was removed from the skill in an earlier revision.

File: .claude/skills/vector-components-maturity-eval/SKILL.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — coverage/lcov phase was removed from the skill in an earlier revision.

File: .claude/skills/vector-components-maturity-eval/SKILL.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Stale — Phase 2c already includes the generated CUE file in the churn path list.

File: .claude/skills/vector-components-maturity-eval/SKILL.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (pront) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Not a meaningful signal — commonly_used is noise, not a reliable proxy for production adoption.

File: .claude/skills/vector-components-maturity-eval/SKILL.md [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+- `minimal`: none meaningful or all are placeholders/references
+
+---
+
+## Phase 4: Match Bugs to Components
+
+For each issue from Phase 2a, match it to components using two signals:
+
+1. **Component labels**: issues are often labeled `source: <name>`, `sink: <name>`, or `transform: <name>`. A label match is authoritative — count the issue for that component without needing a title match.
+2. **Title scan**: scan the title for canonical component names from the CUE filenames. Only count a title match when the component name appears unambiguously (e.g. `"kafka source: ..."`, `"[loki sink]"`, or the name as a standalone token next to "source", "sink", or "transform").
```
[the remark this answers is not in the document]
Comment (thomasqueirozb) [reply .claude/skills/vector-components-maturity-eval/SKILL.md]: Only `sink: splunk_hec` existed but it has now been removed and renamed to `sink: splunk_hec_logs`

**Review by thomasqueirozb — APPROVED** [review]
Nice!
